documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,486 @@
+   UPDATE image
+   SET standardized_popularity = standardized_image_popularity(
+       image.provider,
+       image.meta_data
+   )
+   WHERE identifier in (
+       SELECT identifier FROM rows_to_update
+       WHERE row_id >= i AND row_id < i+rows_per_loop
+       FOR UPDATE SKIP LOCKED
+   );
+   -- COMMIT this batch
+   COMMIT;
+END LOOP;
+END;
+$$;
+```
+
+This is much less efficient than doing a single `UPDATE`, but has clear
+advantages:
+
+- It allows provider DAGs to continue running unimpeded.
+- The data refresh can run at any time, even while the refresh tasks are
+  running.
+- Because updates are committed incrementally, each batch of updated scores will
+  be 'live' on the image table as soon as they are committed. Rather than
+  waiting for all rows to be calculated before making the update, as much
+  up-to-date data as possible will be present each time the data refresh runs.
+
+### Remove popularity steps from the data_refresh DAGs
+
+We can simply remove the popularity tasks from the relevant data refresh DAG
+factories. See the `Parallelizable Streams` section for more information on the
+timing of this change.
+
+### Run the data refresh from the media tables
+
+Once the `image` and `audio` tables contain up-to-date standardized popularity
+data, we can update the data refresh to read from these tables rather than their
+respective materialized views. Because the schema of the media tables now
+matches what the ingestion server already expects, we can simply change the
+table name each place it is referenced:
+
+- `_get_shared_cols`
+- `refresh_api_table`
+- `get_copy_data_query`
+
+## Dependencies
+
+### Tools & packages
+
+<!-- Describe any tools or packages which this work might be dependent on. If multiple options are available, try to list as many as are reasonable with your own recommendation. -->
+
+We should not need any additional dependencies.
+
+### Other projects or work
+
+<!-- Note any projects this plan is dependent on. -->
+
+This does not depend on any existing projects.
+
+### Infrastructure
+
+<!-- Note any infrastructure this plan is dependent on. -->
+
+This project directly affects our Catalog database schema, Airflow DAGs, and
+ingestion server. All necessary changes are described in this document.
+
+Notably, because this plan would allow us to drop the materialized `image_view`
+and `audio_view`, we should net some DB storage savings in the long term.
+
+## Alternatives
+
+### Duplicate table
+
+Another approach was considered which, rather than updating records directly on
+the `image` table, would have instead created a duplicate table with updated
+scores and ultimately swapped tables. This has the benefit of avoiding
+concurrent writes to the `image` table, but the following disadvantages:
+
+- While the duplicate table is being built, writes to the `image` table from
+  provider DAGs would need to be propagated to it via a `Trigger`. Thus the same
```

Comment:
Is this a Postgres `Trigger` or a `Trigger<x>` Operator in Airflow? (Do you mind linking to the documentation here as well?)

Reply:
I was thinking of a [Postgres Trigger](https://www.postgresql.org/docs/current/plpgsql-trigger.html), but it's a really good point that needs to be clarified. I'll add the link to the doc here as well!

FWIW I actually got as far as prototyping this approach before realizing we could do it without fully duplicating the table, so I even have the trigger I _would_ have proposed:

```pgsql
CREATE OR REPLACE FUNCTION refresh_copy()
  RETURNS TRIGGER
  LANGUAGE plpgsql
AS $$
  BEGIN
    IF (TG_OP = 'DELETE') THEN
        DELETE FROM image_copy WHERE OLD.identifier;
     ELSE
        DELETE FROM image_copy where identifier = NEW.identifier;
        INSERT INTO image_copy
          SELECT *
          FROM image
          WHERE identifier = NEW.identifier;
     END IF;
     RETURN NULL;
  END;
$$;

CREATE CONSTRAINT TRIGGER refresh_copy_trigger AFTER INSERT OR DELETE OR UPDATE ON image FOR EACH ROW EXECUTE PROCEDURE refresh_copy();
```